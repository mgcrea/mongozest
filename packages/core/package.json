{
  "name": "@mongozest/core",
  "version": "0.2.8",
  "description": "",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "module": "src/index.ts",
  "scripts": {
    "build": "yarn build:sources && yarn build:typings",
    "build:sources": "babel --source-maps --extensions .js,.ts --out-dir lib/ --delete-dir-on-start --ignore **/__tests__ src/",
    "build:typings": "tsc --declaration",
    "typecheck": "tsc --pretty --noEmit",
    "test": "CI=true jest --runInBand --colors"
  },
  "author": "Olivier Louvignes <olivier@mgcrea.io>",
  "license": "MIT",
  "dependencies": {
    "@babel/runtime": "^7.11.2",
    "@mongozest/hooks": "0.2.8",
    "chalk": "^4.1.0",
    "faker": "^5.1.0",
    "lodash": "^4.17.20",
    "mongodb": "^3.6.2",
    "pluralize": "^8.0.0",
    "shortid": "^2.2.15"
  },
  "devDependencies": {
    "@babel/cli": "^7.11.6",
    "@babel/core": "^7.11.6",
    "@types/jest": "^26.0.14",
    "@types/mongodb": "^3.5.27",
    "@types/pluralize": "^0.0.29",
    "@types/shortid": "^0.0.29",
    "jest": "^26.4.2",
    "typescript": "^4.0.3"
  }
}
