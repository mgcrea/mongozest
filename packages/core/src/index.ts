export * from './utils/objectId';
export * from './utils/logger';

import Model, {OperationMap} from './model';
import {Schema, SchemaProperties, JsonSchema} from './schema';
import jsonSchemaPlugin from './plugins/jsonSchemaPlugin';
import Interface from './interface';

export {Interface, Model, Schema, SchemaProperties, JsonSchema, OperationMap};
export {ObjectId, MongoError} from 'mongodb';

export {mapPathValues, defaultPathValues} from './utils/traversing';
export {jsonSchemaPlugin};

export default Interface.create;
