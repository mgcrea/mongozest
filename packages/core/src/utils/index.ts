export * from './logger';
export * from './objectId';
export * from './traversing';
