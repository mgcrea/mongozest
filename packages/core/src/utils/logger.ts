import chalk from 'chalk';
import {inspect as defaultInspect} from 'util';
import {toString} from 'lodash';

export {chalk};

export const chalkString = (s: string): string => chalk.green(`'${s}'`);
export const chalkStringArray = (a: Array<string>): string => `[ ${a.map(chalkString).join(', ')} ]`;
export const chalkNumber = (n: number | string): string => chalk.yellow(toString(n));
export const chalkBoolean = (b: boolean): string => chalk.yellow(b ? 'true' : 'false');
export const chalkDate = (d: Date): string => chalk.magenta(d.toISOString());

export const inspect = (maybeObject: unknown): string =>
  defaultInspect(maybeObject, {compact: true, colors: true, depth: Infinity, breakLength: Infinity});

export const log = (maybeString: string): void =>
  console.log(`${chalk.gray(new Date().toISOString())} - 🏛 mongodb: ${maybeString}`);
