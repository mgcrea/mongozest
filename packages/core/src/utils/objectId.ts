import {intersectionWith, unionWith, uniqBy, uniqWith} from 'lodash';
import {ObjectId} from 'mongodb';

export const includesObjectId = (arrayOfObjectIds: ObjectId[], targetObjectId: ObjectId) =>
  arrayOfObjectIds.some(objectId => objectId.equals(targetObjectId));

export const intersectionWithObjectIds = (...arraysOfObjectIds: ObjectId[]) =>
  intersectionWith<ObjectId, ObjectId>(...arraysOfObjectIds, (arrVal: ObjectId, othVal: ObjectId) =>
    arrVal.equals(othVal)
  );

export const unionWithObjectIds = (...arraysOfObjectIds: ObjectId[]) =>
  unionWith<ObjectId>(...arraysOfObjectIds, (arrVal: ObjectId, othVal: ObjectId) => arrVal.equals(othVal));

export const uniqWithObjectIds = (arrayOfObjectIds: ObjectId[]) =>
  uniqWith(arrayOfObjectIds, (arrVal, othVal) => arrVal.equals(othVal));

export const uniqByObjectId = (arrayOfObjects: {_id: ObjectId}[]) =>
  uniqBy(arrayOfObjects, arrVal => arrVal._id.toString());
