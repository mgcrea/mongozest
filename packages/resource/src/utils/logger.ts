export { inspect, log } from '@mongozest/core';
