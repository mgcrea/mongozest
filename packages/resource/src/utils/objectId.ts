import {intersectionWith, unionWith, uniqBy, uniqWith} from 'lodash';

export const includesObjectId = (arrayOfObjectIds, targetObjectId) =>
  arrayOfObjectIds.some(objectId => objectId.equals(targetObjectId));
export const intersectionWithObjectIds = (...arraysOfObjectIds) =>
  intersectionWith(...arraysOfObjectIds, (arrVal, othVal) => arrVal.equals(othVal));
export const unionWithObjectIds = (...arraysOfObjectIds) =>
  unionWith(...arraysOfObjectIds, (arrVal, othVal) => arrVal.equals(othVal));
export const uniqWithObjectIds = arrayOfObjectIds =>
  uniqWith(arrayOfObjectIds, (arrVal, othVal) => arrVal.equals(othVal));
export const uniqByObjectId = arrayOfObjects => uniqBy(arrayOfObjects, arrVal => arrVal._id.toString());
