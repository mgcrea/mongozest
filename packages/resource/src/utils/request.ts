export const parseBodyAsUpdate = (body: { [s: string]: any }) =>
  Object.keys(body).reduce(
    (soFar: { [s: string]: any }, key: string) => {
      if (!key.startsWith('$')) {
        soFar.$set[key] = body[key];
      } else {
        soFar[key] = body[key];
      }
      return soFar;
    },
    { $set: {} }
  );

export const asyncHandler = (wrappedFunction: Function) =>
  function asyncWrap(...args: any[]) {
    const next = args[args.length - 1];
    try {
      Promise.resolve(wrappedFunction(...args)).catch(next);
    } catch (err) {
      next(err);
    }
  };
