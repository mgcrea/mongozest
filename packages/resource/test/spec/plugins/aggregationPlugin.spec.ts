import { AggregationPipeline, Model, ObjectId, Schema } from '@mongozest/core';
import createResource, { Resource } from '@mongozest/resource';
import { omit } from 'lodash';
import { getDbName } from 'root/test/utils';
import { makeFetch } from 'supertest-fetch';
import { breakdownMiddleware, createTestApp, fixtures } from '../../utils/';

const DB_NAME = getDbName(__filename);

const app = createTestApp({ routers: [] });
const { mongo, insertFixture } = app.locals;
app.locals.fixtures = fixtures;
const fetch = makeFetch(app);

type User = {
  firstName?: string;
  lastName?: string;
  email: string;
  nationality?: string;
  device?: ObjectId;
};

class UserModel extends Model<User> {
  static modelName = 'User';
  static schema: Schema<User> = {
    firstName: { bsonType: 'string' },
    lastName: { bsonType: 'string' },
    email: { bsonType: 'string', required: true },
    nationality: { bsonType: 'string' },
    device: { bsonType: 'objectId' },
  };
}

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
  await mongo.loadModel(UserModel);
});

afterAll(async () => {
  await Promise.all([app.close()]);
});

describe('Resource', () => {
  let resource: Resource<User>;
  beforeAll(async () => {
    resource = createResource('User');
    expect(resource instanceof Resource).toBeTruthy();
  });
  describe('resource', () => {
    it('should properly serve resource', async () => {
      const router = resource.buildRouter();
      app.use(router);
      app.use(breakdownMiddleware);
    });
  });
  describe('collection', () => {
    describe('GET /users/aggregate', () => {
      it('should return 200', async () => {
        const { insertedId } = await insertFixture('User.mongozest');
        const pipeline: AggregationPipeline = [];
        const query = `pipeline=${JSON.stringify(pipeline)}`;
        const res = await fetch(`/users/aggregate?${query}`, {
          method: 'get',
          headers: { 'Content-Type': 'application/json' },
        })
          .expect(200)
          .expect('content-type', /^application\/json/);
        const resBody = await res.json();
        expect(Array.isArray(resBody)).toBeTruthy();
        expect(resBody.length).toEqual(1);
        expect(omit(resBody[0], '_id')).toMatchSnapshot();
      });
    });
  });
});
