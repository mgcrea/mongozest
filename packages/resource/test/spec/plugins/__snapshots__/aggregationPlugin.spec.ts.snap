// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource collection GET /users/aggregate should return 200 1`] = `
Object {
  "device": "1a1a836f1cf9087b520148de",
  "email": "john.doe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;
